### ISSUE TITLE
Nested dependencies are different between development mode and production mode

### ISSUE THREAD

**Issue by yoriiis** [webpack/webpack#12202]
# Bug report

**What is the current behavior?**

With webpack v5, the following code which produce different results between development mode and production mode. Nested dependencies are not exposed by the compilation in production mode, while in development mode the full list is available.

```js
const path = require('path');

function getSvgsDependenciesByEntrypoint(entryName) {
    let listSvgsDependencies = [];

    compilation.entrypoints.get(entryName).chunks.forEach((chunk) => {
        compilation.chunkGraph.getChunkModules(chunk).forEach((module) => {
            module.dependencies.forEach((dependency) => {
                const extension = path.extname(dependency.userRequest).substr(1);
                if (extension === 'svg') {
                    const moduleDependency = compilation.moduleGraph.getModule(dependency);
                    listSvgsDependencies.push(moduleDependency);
                }
            });
        });
    });

    return listSvgsDependencies;
}
```
Example of a nested dependency tree with an entry point `app-a`

```
- app-a.js
    - svg1.svg
    - module1.js
        - svg2.svg
        - svg3.svg
        - module2.js
            - svg4.svg
            - svg5.svg
```

In development build

```js
const svgs = getSvgsDependenciesByEntrypoint('app-a');
console.log(svgs) // ["svg1.svg", "svg2.svg", "svg3.svg", "svg4.svg", "svg5.svg"]
```

In production build

```js
const svgs = getSvgsDependenciesByEntrypoint('app-a');
console.log(svgs) // ["svg1.svg"]
```

**If the current behavior is a bug, please provide the steps to reproduce.**

1. Clone the code from `https://github.com/yoriiis/svg-chunk-webpack-plugin/pull/1`
2. Install dependencies with `npm install`
3. Run the example in development mode with `npm run start:example`
4. Open the `example/dist/sprites-manifest.json` file
5. SVG files for the `home` entry point are `gradient.svg`, `video.svg` and `smiley-love.svg`
6. Run the example in production mode with `npm run build:example`
7. Open the `example/dist/sprites-manifest.json` file
8. SVG files for the `home` entry point are `gradient.svg` and `video.svg`

**What is the expected behavior?**

The `home` entry point does not list the nested SVG files.
The plugin `svg-chunk-webpack-plugin` used the compilation object and needs a flat list of file dependencies filtered by entry point to extract all SVGs for each entry points.

**Other relevant information:**
webpack version:  5.4.0
Node.js version: 10.13.0
Operating System: macOS Catalina

I've also ask the question on StackOverflow:
https://stackoverflow.com/questions/64903453/how-can-i-get-a-flat-list-of-file-dependencies-filtered-by-entry-point-from-the

**Comment by sokra** [comment]
This is working as expected. In your example none of your modules have side effects, webpack 5 is able to detect that and omit all modules from the bundle. The output JS files are also empty.

[v5.10.3](https://github.com/webpack/webpack/releases/tag/v5.10.3) includes a fix which allows to manually specify the a module has side effects. You could do that either in module rules or within the loader directly.

``` js
		module: {
			rules: [
				{
					test: /\.svg$/,
					use: [
						{
							loader: SvgChunkWebpackPlugin.loader
						}
					],
					sideEffects: true
				}
			]
		},
```

``` js
export = function spriteLoader(this: any, content: string): string {
	// ...
	this._module.factoryMeta = this._module.factoryMeta || {};
	this._module.factoryMeta.sideEffectFree = false;
};
```

That should help with that problem.

---

btw. the way of iterating the modules is a bit weird, you don't have to use the `dependencies` to find the svgs.

It would do it this way:

``` js
	getSvgsDependenciesByEntrypoint(entryName: string): Array<NormalModule> {
		let listSvgsDependencies: Array<NormalModule> = [];

		this.compilation.entrypoints.get(entryName).chunks.forEach((chunk: any) => {
			for (const module of this.compilation.chunkGraph.getChunkModulesIterable(chunk)) {
				if (module.buildInfo && module.buildInfo.SVG_CHUNK_WEBPACK_PLUGIN) {
					listSvgsDependencies.push(module);
				}
			}
		});

		return listSvgsDependencies;
	}
```

and flag svg processed with your loader:

``` js
export = function spriteLoader(this: any, content: string): string {
	// ...
	this._module.buildInfo.SVG_CHUNK_WEBPACK_PLUGIN = true;
};
```

---

I think it would make even more sense when keeping the existing semantics for importing assets:

``` js
import url from "./thing.svg"
```

So importing the SVG give an URL to the image. In your case it could be something like `data:image/svg,<svg><use xlink:href="#<SVG_FILENAME>"></use></svg>`

---

I also think it would make more sense to generate a spritemap per chunk instead of per entrypoint...

**Comment by yoriiis** [comment]
Thanks @sokra for the quick reply!

I did not know this part of side effects 👍 
The JS files are empty because the SVG content is exported in a dedicated file, so this is the normal behavior for my need.
The two solutions for the side effects declarations works perfectly. I prefer the one with the declaration inside the loader to avoid conflicts.

With the dependent of the release `v5.10.3`, I lost webpack v4/v5 compatibility. Any solution for this?

With webpack 4, I didn't need this option, why? (Without this option, nested dependencies are identical in both mode).

---

Thanks for the update of the `getSvgsDependenciesByEntrypoint` function with the flag, it's better and more efficient!

---

For the syntax `import url from "./thing.svg"`, the loader exports the content of the SVG without modification, so that the plugin can use it. And, should the variable `url` return the `use` for the sprite or directly the `<svg>`? Sorry I don't see how this part could work🤔

```js
export = function spriteLoader(this: any, content: string): string {
	// ...
	return JSON.stringify(content);
};

```js
import url from "./thing.svg"
```

---

The plugin generates a SVG sprite with all SVG files used by an entry point. The current spritemap reflects this principle.
An entry point can contains several chunks, what would be the advantage with the "per chunk" cutting?

**Comment by sokra** [comment]
> The plugin generates a SVG sprite with all SVG files used by an entry point. The current spritemap reflects this principle.
> An entry point can contains several chunks, what would be the advantage with the "per chunk" cutting?

Multiple entrypoints would share a spritemap. Basically you would get a "commons" spritemap with the shared svg and an additional spritemap with the svgs only used by this entrypoint. Similar how JS code is shared and splitted.

Per chunk spritemaps would also allow on demand loading of spritted SVGs, but the additional spritemaps also need to be loaded somehow, so that becomes complex.

But this is optional and if you don't want to that's fine.

> Sorry I don't see how this part could work🤔

I'm also not sure exactly. I was thinking of using an URL like `data:image/svg,<svg><use xlink:href="#<SVG_FILENAME>"></use></svg>` which is an inline SVG containing only: 

``` svg
<svg>
  <use xlink:href="#<SVG_FILENAME>"></use>
</svg>
```

I'm not that experienced with SVGs, maybe that doesn't work.

Even cooler would be if that would work:

``` svg
<svg>
  <use xlink:href="<SVG_SPRITEMAP>#<SVG_FILENAME>"></use>
</svg>
```

**Comment by yoriiis** [comment]
Sorry for the delay @sokra, this issue is resolved for me, the release `v5.10.3` fixes the bug.
I am working on the migration to finalize compatiblity with webpack v5. Finally there will not be the backward compatibility with v4, but I think it's not essential (the new version will be major).

Regarding the proposal of shared and split code in SVG as in the JS, it is another feature, I'll think about it, thank you for the idea.
